ext/dom/xpath.c

```diff
@@ -446,6 +446,60 @@ PHP_METHOD(DOMXPath, registerPhpFunctionNS)
 	);
 }
 
+/* {{{ */
+PHP_METHOD(DOMXPath, quote) {
+	char *input;
+	size_t input_len;
+	smart_str output = {0};
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "s", &input, &input_len) ==
+			FAILURE) {
+		RETURN_THROWS();
+	}
+	if (memchr(input, '\'', input_len) == NULL) {
+		smart_str_appendc(&output, '\'');
```

Comment:
Technically, you could do it more performant for the cases where only a single quote type (i.e. either only ' or ") is used by:
* Allocate zend_string using zend_string_safe_alloc with precomputed length.
* Do memcpy + manual character setting
* Then return that string using  ZVAL_NEW_STR(return_value, ...)

That's a little more optimal than using smart_str for these two branches because there are no reallocations then in that case. It does make sense to use smart_str in the complex else branch.
Anyway, I don't object to doing it via smart_str though.

Reply:
Sounds good, thanks.

Btw the original implementation was entirely with manual memory management/emalloc/memcpy: 1730af162a66aee420cdda35f0359188ec0ce859